test_conformance/subgroups/workgroup_common_kernels.cpp

```diff
@@ -0,0 +1,104 @@
+//
+// Copyright (c) 2020 The Khronos Group Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include "workgroup_common_kernels.h"
+
+const char* bcast_source = "__kernel void test_bcast(const __global Type *in, "
+                           "__global int4 *xy, __global Type *out)\n"
+                           "{\n"
+                           "    int gid = get_global_id(0);\n"
+                           "    XY(xy,gid);\n"
+                           "    Type x = in[gid];\n"
+                           "    out[gid] = sub_group_broadcast(x, xy[gid].z);\n"
+
+                           "}\n";
+
+const char* redadd_source = "__kernel void test_redadd(const __global Type "
+                            "*in, __global int4 *xy, __global Type *out)\n"
+                            "{\n"
+                            "    int gid = get_global_id(0);\n"
+                            "    XY(xy,gid);\n"
+                            "    out[gid] = sub_group_reduce_add(in[gid]);\n"
+                            "}\n";
+
+const char* redmax_source = "__kernel void test_redmax(const __global Type "
+                            "*in, __global int4 *xy, __global Type *out)\n"
+                            "{\n"
+                            "    int gid = get_global_id(0);\n"
+                            "    XY(xy,gid);\n"
+                            "    out[gid] = sub_group_reduce_max(in[gid]);\n"
+                            "}\n";
+
+const char* redmin_source = "__kernel void test_redmin(const __global Type "
+                            "*in, __global int4 *xy, __global Type *out)\n"
+                            "{\n"
+                            "    int gid = get_global_id(0);\n"
+                            "    XY(xy,gid);\n"
+                            "    out[gid] = sub_group_reduce_min(in[gid]);\n"
+                            "}\n";
+
+const char* scinadd_source =
+    "__kernel void test_scinadd(const __global Type *in, __global int4 *xy, "
+    "__global Type *out)\n"
+    "{\n"
+    "    int gid = get_global_id(0);\n"
+    "    XY(xy,gid);\n"
+    "    out[gid] = sub_group_scan_inclusive_add(in[gid]);\n"
```

Comment:
I think this was brought up before.
Are there any provisions to ensure that implementations are using the correct overload?
That is, what's stopping an implementation from type-promoting "char" to "int" and using the "int" overload of this function?

Reply:
maybe such check should work?  `sizeof(int[gid]) == sizeof(sub_group_scan_inclusive_add(in[gid]))`

Reply:
Hm, that might.

Reply:
Example  [commit](https://github.com/KhronosGroup/OpenCL-CTS/pull/1074/commits/65cd0ad1ea6f8adc3344417f3d04892e98b933f6) for cluster reduce function has been added. Test runs ok. If you are ok with that I can add such solution for other functions.

work_group_functions_clustered_reduce...
  sub_group_clustered_reduce_add(int, 4 bytes) ...
  sub_group_clustered_reduce_max(int, 4 bytes) ...
  sub_group_clustered_reduce_min(int, 4 bytes) ...
  sub_group_clustered_reduce_mul(int, 4 bytes) ...
  sub_group_clustered_reduce_and(int, 4 bytes) ...
  sub_group_clustered_reduce_or(int, 4 bytes) ...
  sub_group_clustered_reduce_xor(int, 4 bytes) ...
  sub_group_clustered_reduce_add(uint, 4 bytes) ...
  sub_group_clustered_reduce_max(uint, 4 bytes) ...
  sub_group_clustered_reduce_min(uint, 4 bytes) ...
  sub_group_clustered_reduce_mul(uint, 4 bytes) ...
  sub_group_clustered_reduce_and(uint, 4 bytes) ...
  sub_group_clustered_reduce_or(uint, 4 bytes) ...
  sub_group_clustered_reduce_xor(uint, 4 bytes) ...
  sub_group_clustered_reduce_add(long, 8 bytes) ...
  sub_group_clustered_reduce_max(long, 8 bytes) ...
  sub_group_clustered_reduce_min(long, 8 bytes) ...
  sub_group_clustered_reduce_mul(long, 8 bytes) ...
  sub_group_clustered_reduce_and(long, 8 bytes) ...
  sub_group_clustered_reduce_or(long, 8 bytes) ...
  sub_group_clustered_reduce_xor(long, 8 bytes) ...
  sub_group_clustered_reduce_add(ulong, 8 bytes) ...
  sub_group_clustered_reduce_max(ulong, 8 bytes) ...
  sub_group_clustered_reduce_min(ulong, 8 bytes) ...
  sub_group_clustered_reduce_mul(ulong, 8 bytes) ...
  sub_group_clustered_reduce_and(ulong, 8 bytes) ...
  sub_group_clustered_reduce_or(ulong, 8 bytes) ...
  sub_group_clustered_reduce_xor(ulong, 8 bytes) ...
  sub_group_clustered_reduce_add(short, 2 bytes) ...
  sub_group_clustered_reduce_max(short, 2 bytes) ...
  sub_group_clustered_reduce_min(short, 2 bytes) ...
  sub_group_clustered_reduce_mul(short, 2 bytes) ...
  sub_group_clustered_reduce_and(short, 2 bytes) ...
  sub_group_clustered_reduce_or(short, 2 bytes) ...
  sub_group_clustered_reduce_xor(short, 2 bytes) ...
  sub_group_clustered_reduce_add(ushort, 2 bytes) ...
  sub_group_clustered_reduce_max(ushort, 2 bytes) ...
  sub_group_clustered_reduce_min(ushort, 2 bytes) ...
  sub_group_clustered_reduce_mul(ushort, 2 bytes) ...
  sub_group_clustered_reduce_and(ushort, 2 bytes) ...
  sub_group_clustered_reduce_or(ushort, 2 bytes) ...
  sub_group_clustered_reduce_xor(ushort, 2 bytes) ...
  sub_group_clustered_reduce_add(uchar, 1 bytes) ...
  sub_group_clustered_reduce_max(uchar, 1 bytes) ...
  sub_group_clustered_reduce_min(uchar, 1 bytes) ...
  sub_group_clustered_reduce_mul(uchar, 1 bytes) ...
  sub_group_clustered_reduce_and(uchar, 1 bytes) ...
  sub_group_clustered_reduce_or(uchar, 1 bytes) ...
  sub_group_clustered_reduce_xor(uchar, 1 bytes) ...